pulp_container/app/viewsets.py

```diff
@@ -119,7 +124,99 @@ class Meta:
+    "container.container",
+)
+
+
+class ContainerContentQuerySetMixin:
+    """
+    A mixin that provides container content models with querying utilities.
+    """
+
+    def _repo_query_params(self, request, view, repo_info):
+        """
+        Checks if the requests' query_params contain repository_version.
+
+        This is used in the quryset scoping for content.
+
+        Args:
+            request (rest_framework.request.Request): The request being made.
+            view (subclass rest_framework.viewsets.GenericViewSet): The view being checked for
+                authorization.
+            action (str): The action being performed, e.g. "destroy".
+            repo_info (tuple): Tuple that contains repo type and  Permissions to
+                be checked.
+
+        Returns:
+            List of repositories pk that the current user can view
+
+        """
+        repo_pks = []
+        for key, param in request.query_params.items():
+            if "repository_version" in key:
+                rv = view.get_resource(param, RepositoryVersion)
+                repo = rv.repository.cast()
+                if (
+                    request.user.has_perm(repo_info.push_perm)
+                    and repo.pulp_type == repo_info.push_type
+                ):
+                    repo_pks.append(repo.pk)
+                elif (
+                    request.user.has_perm(repo_info.mirror_perm)
+                    and repo.pulp_type == repo_info.mirror_type
+                ):
+                    repo_pks.append(repo.pk)
+                elif request.user.has_perm(repo_info.push_perm, repo) or request.user.has_perm(
+                    repo_info.mirror_perm, repo
+                ):
+                    repo_pks.append(repo.pk)
+        return repo_pks
+
+    def get_queryset(self):
+        """
+        Gets a QuerySet based on the current request.
+
+        Filters and retuns the only the repo's content user is allowed to see.
+
+        Returns:
+            django.db.models.query.QuerySet: The queryset returned contains content the user is
+            allowed to see based on the repo permissions.
+
+        """
+        qs = super().get_queryset()
+        has_model_push_repo = self.request.user.has_perm(repo_info.push_perm)
+        has_model_repo = self.request.user.has_perm(repo_info.mirror_perm)
+        # this will show also orphaned content
+        if has_model_push_repo and has_model_repo:
+            return qs
+        query_params = self.request.query_params
+        if query_params and "repository_version" in query_params:
+            repo_pks = self._repo_query_params(self.request, self, repo_info)
+            content_qs = qs.model.objects.filter(repositories__in=repo_pks)
+        else:
+            allowed_push_repos = get_objects_for_user(
+                self.request.user,
+                repo_info.push_perm,
+                klass=models.ContainerPushRepository.objects.all(),
```

Comment:
Did you add this because specifying just model didn't work?
I'm again unsure about queryset evaluation, I think that all() might cause it to be executed right away while filtering by model is potentially more optimized.
I read here that you can specify a model https://django-guardian.readthedocs.io/en/stable/api/guardian.shortcuts.html#get-objects-for-user
```
klass – may be a Model, Manager or QuerySet object. If not given this parameter would be computed based on given params.
```

Reply:
I had an impression klass accepts only a queryset that;s why i used it. Let me try with the model